MapleServer2/PacketHandlers/Game/InsigniaHandler.cs

```diff
@@ -14,15 +17,80 @@ public class InsigniaHandler : GamePacketHandler
 
         public override void Handle(GameSession session, PacketReader packet)
         {
+
+            ProcessInsignias(session);
+
             short insigniaId = packet.ReadShort();
 
             if (insigniaId < 0)
             {
                 return;
             }
 
-            session.Player.InsigniaId = insigniaId;
-            session.FieldManager.BroadcastPacket(InsigniaPacket.UpdateInsignia(session, insigniaId, true));
+            if (session.Player.AvaliableInsignias.Contains(insigniaId))
+            {
+                session.Player.InsigniaId = insigniaId;
+            }
+            else
+            {
+                insigniaId = 0;
+            }
+
+            session.FieldManager.BroadcastPacket(InsigniaPacket.UpdateInsignia(session, insigniaId, insigniaId != 0));
+        }
+
+        private void ProcessInsignias(GameSession session)
+        {
+            Player player = session.Player;
+            int[] trophy = player.Trophy;
+            List<int> avaliableTitles = player.AvaliableTitles;
+            List<short> tempInsignias = new List<short> { };
+            Dictionary<ItemSlot, Item> equips = player.Equips;
+
+            Dictionary<int, short> titlesThatGiveInsignias = new Dictionary<int, short> {
+                {10000569, 2}, {10000152, 3}, {10000570, 4}, {10000196, 6}, {10000195, 7}, {10000571, 14}, {10000331, 15}, {10000190, 22},
+                {10000458, 28}, {10000465, 30 }, {10000503, 33}, {10000512, 35}, {10000513, 36}, {10000514, 37}, {10000537, 38}, {10000565, 41},
+                {10000602, 43}, {10000603, 44}, {10000638, 45 }, {10000644, 46}
+            };
+
+            // TODO: Missing checks for Mushking Royale lv 5 and Premium Club member
+            int trophyCount = trophy[0] + trophy[1] + trophy[2];
+
+            if (avaliableTitles.Contains(10000171) && trophyCount >= 1000 && !tempInsignias.Contains(5))
```

Comment:
Instead of always calculating, maybe we should just save and load what insignia someone has. There's tradeoffs for both I guess. However, later on when you get insignias you can purchase and are not limited to trophies, having a record of what insignias someone has might work better.

Reply:
Yes having a record of what insignias someone has is the better way to deal but some of them are not granted forever and rely on variables like if the player is in a premium club or have an equip +12 equipped. These are the edges cases I'm aware that needs to be checked every time someone tries to equip an insignia.